### PR TITLE
console - log relevant information in home, user and logs page

### PR DESCRIPTION
**Pull request by Gaetanbrl**
Link to #2267 .

Display logs detail by click to get log relevant information in admin log.

Thank you in advance for review.
@fvanderbiest @landryb @tonio @RemiDesgrange

### CODE REVIEW

**Comment by landryb** [comment]
Can we take https://github.com/georchestra/georchestra/issues/2267#issuecomment-542096071 into account ? once the `changed` field is added to the schema, i think it's hard to migrate its type/size to something else, so might aswell get it large from the start.

Apparently (per  https://thoughts-on-java.org/mapping-blobs-and-clobs-with-hibernate-and-jpa/)there are two ways, using the `Clob` java type instead of `String`, or just annotate the property and keep using `String`.

I tried adding `@Lob`  annotion on the `changed` definition, dropped the `changed` column in the db with `varchar` type, it was recreated by hibernate as `text` type in the DB - modifying an attr on a user works (ie json contains `{"new":"editoraaa","field":"description","old":"editor","type":"Attribute was changed for a user"}`) but strangely the json is not directly readable in the database field (there's `733070` in the field for that record ?). - and here i will stop poking at this since im no java hell expert.

**Comment by Gaetanbrl** [comment]
> Can we take #2267 (comment) into account ?

Unfortunately, we can't take it into account now. I can open a new issue to keep this in mind.

> I tried adding @Lob annotion on the changed definition

Thank you,  hope we can do that soon...

**Comment by landryb** [comment]
Ahem... gentle review poke ? :)

**Comment by Gaetanbrl** [comment]
need to manage conflict / rebase again ...

**Comment by RemiDesgrange** [comment]
Can you make the integration test pass ? Thanks.

**Comment by Gaetanbrl** [comment]
> Can you make the integration test pass ? Thanks.

Yes, we can @RemiDesgrange.  As said before : 
> need to manage conflict / rebase

**Comment by pierrejego** [comment]
There are three integration tests on attributes deletion for orgs which failed for the moment. I am trying to find out why. 
We can delete those attributes in real GUI, without errors.

**Comment by pierrejego** [comment]
> Can we take [#2267 (comment)](https://github.com/georchestra/georchestra/issues/2267#issuecomment-542096071) into account ? once the `changed` field is added to the schema, i think it's hard to migrate its type/size to something else, so might aswell get it large from the start.
> 
> Apparently (per https://thoughts-on-java.org/mapping-blobs-and-clobs-with-hibernate-and-jpa/)there are two ways, using the `Clob` java type instead of `String`, or just annotate the property and keep using `String`.
> 
> I tried adding `@Lob` annotion on the `changed` definition, dropped the `changed` column in the db with `varchar` type, it was recreated by hibernate as `text` type in the DB - modifying an attr on a user works (ie json contains `{"new":"editoraaa","field":"description","old":"editor","type":"Attribute was changed for a user"}`) but strangely the json is not directly readable in the database field (there's `733070` in the field for that record ?). - and here i will stop poking at this since im no java hell expert.

Last error in integration test was due to attributed changed size. (char varying 255) to small to set all information. If an error occurs when trying to save in database, en Error log is provided in trace, and changed value contains. JSon "error": "Error while inserting admin log in database, see admin log file"

I agree with you @landryb , we should change to Lob, but I am not sure how to do it correctly.

I have added @Lob annotation to "changed" attribut from class "AdminLogEntry", this work perfectly for me; 
I saw that   <prop key="hibernate.hbm2ddl.auto">update</prop> is set in webmvc-config.xml so table should be updated at startup , but It could be complicated to reinstall previous version if needed.

Tell me if I need to do something else to take type change to Log into account.

**Comment by pierrejego** [comment]
I need to make more test before review

**Comment by pierrejego** [comment]
Ok I have done more test.
 1 - New instance OK
 2 - migration from exisiting instance KO,
        hibernate does not change type automaticlly if there is date in it (not on my server), so we will need to Drop previous table, OR make an manual ALTER TABLE Script before migration.  Where should I put this type of script from version migration ?

Otherwise it's ready for review @RemiDesgrange :) thanks

**Comment by landryb** [comment]
as discussed over the phone we dont need to care about the varchar->lob migration since the `changed` field is a new field :)

**Comment by RemiDesgrange** [comment]
The migration should go [here](https://github.com/georchestra/georchestra/tree/master/postgresql) (but ping @pmauduit anyway), you should not use `Lob` to store json in database. Georchestra didn't choose the best opensource database out there to end up usng the DB like if it was MySQL :-) . I am _not_ blaming you for this choice. But using lob (so storing binary) is the not a good idea, because we won't be able to request the JSON from SQL. So every people that build tools around georchestra won't able to just query the data and use it. With JSON type you can. If using JSONB, you can even put indexes on json keys ! 


Ok solution time: you should pass a `JSONObject` to `AdminLogEntry`. 
https://github.com/georchestra/georchestra/blob/f348cc2e42a680a2cca0f753cc0e1090b3c7adbc/console/src/main/java/org/georchestra/console/model/AdminLogEntry.java#L48

This will allow you to remove codes that convert back and forth from `String` to `JSON`

**Comment by fvanderbiest** [comment]
> The migration should go [here](https://github.com/georchestra/georchestra/tree/master/postgresql)

Rather https://github.com/georchestra/georchestra/tree/master/migrations/19.12 (the directory does not exist yet, feel free to create it)

**Comment by landryb** [comment]
Nice, didnt know that json & jsonb were native postgresql data types.

As for the migration, i'm still not sure there should be one, as we're only adding a new field to an existing table, and from my understanding hibernate will create the field if it doesnt exist, so there should be no need for an sql snippet ?

**Comment by tonio** [comment]
_For the frontend side:_

## Architecture

 - There is *a lot* of code duplication. The whole log table should be a shared configurable component, instantiated in the 3 places it is used (It's possible to get inspiration from the `stats` component, shared in home page & stats tabs).

 - Moreover, having a seperate template would make it more readable. There's also too much logic in  the template that should be moved to the component.

## UX

 - There's in my opinion a suboptimal use of available space: last column with change token is not very useful, for most of the logs, details can be displayed directly 
Screenshot from 2019-11-29 11-00-11

 - The breadcrumb which only displayed on item opening is quite confusing (hidden on global view). I think user shouldn't have to leave the list to see log detail. For the longest (ie: message), it could be displayed in a tooltip.
Screenshot from 2019-11-29 11-24-52

## Data

 - Editing multiples fields at once should end in a single log entry, otherwise logs are spammed with quite duplicated entries:
Screenshot from 2019-11-29 13-20-02


## Conclusion

There's at least a refactoring to share the component to be done before going deeper in the review. UX can be discussed but should be simplified in my point of view.

**Comment by landryb** [comment]
> ## UX
> 
>     * There's in my opinion a suboptimal use of available space: last column with change token is not very useful, for most of the logs, details can be displayed directly
>       Screenshot from 2019-11-29 11-00-11

right now we have the 'correct' amount of details given the available space in most cases, can you detail examples you have in mind where the displayed info is not useful enough ?

> ## Data
> 
>     * Editing multiples fields at once should end in a single log entry, otherwise logs are spammed with quite duplicated entries:
>       Screenshot from 2019-11-29 13-20-02

That was a design choice in conjunction with what happens in the backend. IIrc it was a bit too complicated to "merge" changes happening at the same time, but if it needs to be revisited then so be it. How would you display all user-related changes in a single line ?

**Comment by landryb** [comment]
what's the status of the PR now ? being rewritten ? reviews adressed ?

**Comment by pierrejego** [comment]
> The migration should go [here](https://github.com/georchestra/georchestra/tree/master/postgresql) (but ping @pmauduit anyway), you should not use `Lob` to store json in database. Georchestra didn't choose the best opensource database out there to end up usng the DB like if it was MySQL :-) . I am _not_ blaming you for this choice. But using lob (so storing binary) is the not a good idea, because we won't be able to request the JSON from SQL. So every people that build tools around georchestra won't able to just query the data and use it. With JSON type you can. If using JSONB, you can even put indexes on json keys !
> 
> Ok solution time: you should pass a `JSONObject` to `AdminLogEntry`.
> https://github.com/georchestra/georchestra/blob/f348cc2e42a680a2cca0f753cc0e1090b3c7adbc/console/src/main/java/org/georchestra/console/model/AdminLogEntry.java#L48
> 
> This will allow you to remove codes that convert back and forth from `String` to `JSON`

Hello, there are no simple way to store Java JsonObject to JsonB postgresql type. 
Hibernate does not have OOB support for jsonb type in postgres. 

What is the best option :

1 - do we use an external librairie ? Like this example https://vladmihalcea.com/how-to-map-json-objects-using-generic-hibernate-types/ 

2 -Or to we keep @Lob like it is already existing in console  https://github.com/georchestra/georchestra/blob/f879c98a3b535a5ce071a6327aec2f8f369856f1/console/src/main/java/org/georchestra/console/model/Attachment.java#L42 but only change type to JsonObject to at least avoid conversion.

3 - Or perhaps a another idea ?

Regards

**Comment by RemiDesgrange** [comment]
ORM :roll_eyes: . A PITA as always... I am not against adding a new dependency. It seems well maintained and SO have topics about it. So I am in favor of adding `hibernate-types` to console. It doesn't seem _that_ complicated (but maybe I am wrong).

**Comment by Gaetanbrl** [comment]
> So I am in favor of adding `hibernate-types` to console.

Done. This seems to work fine ! Thanks for your response @RemiDesgrange.

https://vladmihalcea.com/map-string-jpa-property-json-column-hibernate/
https://github.com/vladmihalcea/hibernate-types

**Comment by Gaetanbrl** [comment]
To be more explicit, we cannot directly send `json`  to Postgres `jsonb` column with `hibernate-types`: 
```
    @Type(type = "jsonb")
    @Column(columnDefinition = "jsonb")
    private JSONObject changed;
```

According to _hibernates-types_, We need to use `String` first that will be converted to `jsonb` . This solution allow us to send informations to `jsonb ` column.

```
    @Type(type = "jsonb")
    @Column(columnDefinition = "jsonb")
    private String changed;
```

BTW, we need to keep `String` format (and toString() methods) with some of `json` object in order to send data to `jsonb` type column.

**Comment by Gaetanbrl** [comment]
**Front**
- [ ] Home, user, logs templates to a "logger" component (no dupplication, move logic from template to component)
- [ ] Most uniform code for chosen list insertion (logs tab)

**Back**
- [ ] Change private nature if necessary
- [ ] Check dead code
- [ ] Move logs creation after update or modification
- [ ] Save log changes information as jsonb instead of @Lob
- [ ] Move create logs into controller instead of dao
- [ ] Move some methods from controller to logUtils class

**Comment by landryb** [comment]
rebuilt war/deb from the tip of the branch, deployed - the `changed` column is correctly added to `console.admin_log` table (with type jsonb) by the hibernate goo.

Modified address on an org, and an entry is correctly added to the table:
```
georchestra=> select * from console.admin_log where changed is not null;
 id  | admin  |          date           |      target      | type |                                                  changed
-----+--------+-------------------------+------------------+------+-----------------------------------------------------------------------------------------------------------
 367 | breuil | 2020-01-14 08:08:31.566 | COM_RUY_MONTCEAU |    8 | {"new": "Nowhere blah", "old": "Nowhere", "type": "Attribute was changed for an org", "field": "address"}
```

will do more testing today to make sure all user-facing cases that come to my mind are covered - i cant/wont comment on the code itself, but @pierrejego told me and @fvanderbiest privately this was ready for review. @tonio @RemiDesgrange ?

**Comment by landryb** [comment]
i have to agree that all those `if(logUtils != null)` tests look awkward.. are there really cases where it can happen to be null ?

**Comment by landryb** [comment]
as a regular user, trying to create a new account with a new org creates entries in `ou=pendingusers` & `ou=pendingorgs` LDAP branches but nothing is added to the `admin_log` table.  The bits added in https://github.com/georchestra/georchestra/pull/2777/files#diff-1b00ad33140347dc576f731420bbcf6c dont work for some reason ?

**Comment by Gaetanbrl** [comment]
> but nothing is added to the admin_log table

Indeed, text is displayed correctly in the console but nothing is added to the `admin_log `table. I'll control that. Thanks.

**Comment by landryb** [comment]
Bah, disregard my previous comment, my sql query was only showing entries where changed wasnt NULL, and for those it was the case. Doing more testing..

**Comment by Gaetanbrl** [comment]
>  my sql query was only showing entries where changed wasnt NULL

I could display something if you prefer but after control this part seems to work as expected.

**Comment by pierrejego** [comment]
> i have to agree that all those `if(logUtils != null)` tests look awkward.. are there really cases where it can happen to be null ?

Yep I am on it.

**Comment by landryb** [comment]
* [x] create new pending org -> PENDING_ORG_CREATED logged
* [x] create new pending user -> PENDING_USER_CREATED logged
* [x] modify org address -> ORG_ATTRIBUTE_CHANGED(address) logged
* [x] remove pending org -> PENDING_ORG_REFUSED logged
* [x] remove pending user -> PENDING_USER_REFUSED logged
* [x] validate pending org -> PENDING_ORG_ACCEPTED logged
* [x] validate pending user -> PENDING_USER_ACCEPTED logged
* [ ] user modifiers his profile via `console/account/userdetails`, setting telephone number which wasnt set -> USER_ATTRIBUTE_CHANGED logged for title, facsimileTelephoneNumber and telephoneNumber (wtf?):
```
 374 | abb           | 2020-01-14 09:35:42.324 | abb                      |   21 | {"new": "0123456789", "old": "", "type": "Attribute was changed for a user", "field": "telephoneNumber"}
 375 | abb           | 2020-01-14 09:35:42.335 | abb                      |   21 | {"new": "", "old": "", "type": "Attribute was changed for a user", "field": "facsimileTelephoneNumber"}
 376 | abb           | 2020-01-14 09:35:42.338 | abb                      |   21 | {"new": "", "old": "", "type": "Attribute was changed for a user", "field": "title"}

``` 
+ NPE displayed to the user (ouch) :
```
2020-01-14 09:35:42 GlobalExceptionHandler [ERROR] java.lang.NullPointerException
        at java.lang.String.contentEquals(String.java:1066)
        at org.georchestra.console.ws.edituserdetails.EditUserDetailsFormController.logChanges(EditUserDetailsFormController.java:251)
        at org.georchestra.console.ws.edituserdetails.EditUserDetailsFormController.modify(EditUserDetailsFormController.java:272)
        at org.georchestra.console.ws.edituserdetails.EditUserDetailsFormController.edit(EditUserDetailsFormController.java:194)
        at org.georchestra.console.ws.edituserdetails.EditUserDetailsFormController$$FastClassBySpringCGLIB$$26f36d5a.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:667)
        at org.georchestra.console.ws.edituserdetails.EditUserDetailsFormController$$EnhancerBySpringCGLIB$$fdb16c39.edit(<generated>)
```
* [x] admin modifies user description -> USER_ATTRIBUTE_CHANGED(description) logged
* [x] admin adds user to org -> USER_ATTRIBUTE_CHANGED (org) with said user as target, details show ` Replace the value COM_MEILLARD by stestorg. ` (user was previously linked to `COM_MEILLARD`)
* [x] admin removes user from org -> USER_ATTRIBUTE_CHANGED (org) with said user as target, details show ` Replace the value stestorg by . `
* [x] admin adds user to role DRIVE ->  CUSTOM_ROLE_ADDED (DRIVE)  logged
* [x] admin removes user from role DRIVE ->  CUSTOM_ROLE_REMOVED (DRIVE)  logged
* [x] admin creates role TEST -> ROLE_CREATED logged
* [x] admin deletes role TEST -> ROLE_DELETED logged
* [x] admin directly creates org -> ORG_CREATED logged
* [x] admin deletes org -> ORG_DELETED logged
* [x] admin directly creates user -> USER_CREATED logged
* [x] admin directly deletes user -> USER_DELETED logged
* [x] user modifiers his password via `/console/account/changePassword` -> USER_PASSWORD_CHANGED logged
* [ ] admin modifies org area, changing from 1 to complete dept -> ORG_ATTRIBUTE_CHANGED (cities)  logged, details show ` Replace the value by . ` -> should show amount of comms in dept ? here's the entry in db:
```
 389 | breuil        | 2020-01-14 09:49:13.628 | COM_MAGNET               |    8 | {"new": "", "old": "", "type": "Attribute was changed for an org", "added": 316, "field": "cities", "removed": 0}
```
* [ ] admin modifies org area, changing from complete dept to 1 -> ORG_ATTRIBUTE_CHANGED (cities)  logged, details show ` Replace the value by . ` -> should show 1 entry ? here's the entry in db:
```
 390 | breuil        | 2020-01-14 09:51:04.896 | COM_MAGNET               |    8 | {"new": "", "old": "", "type": "Attribute was changed for an org", "added": 0, "field": "cities", "removed": 316}
```

So most cases work, but some (org area modification, user modifiers his profile) definitely still need runtime testing/fixing.

**Comment by landryb** [comment]
All filters on `/console/manager/logs` do work (woooo! - this page was so useless before....) - and the logs tab on a user page also works, showing modifications for this user account. Yay!

Now i realize that there's no such logs tab for an org page, but oh well, that'll be another issue/feature request/PR... out of scope for this one :)

File: console/pom.xml [hunk 1]
```diff
@@ -375,6 +375,11 @@
       <artifactId>ldaptive</artifactId>
       <version>1.0.3</version>
     </dependency>
+	<dependency>
+	  <groupId>com.vladmihalcea</groupId>
+	  <artifactId>hibernate-types-52</artifactId>
+	  <version>2.8.0</version>
+	</dependency>
```
Comment (RemiDesgrange) [inline console/pom.xml]: Not a bloker but when staring the server, it display  a big message telling us that we should use hypersistence optimizer, maybe there is a way to disable this message.

File: console/src/main/java/org/georchestra/console/model/AdminLogEntry.java [hunk 2]
```diff
@@ -44,6 +50,10 @@
     private AdminLogType type;
     private static final DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
 
+    @Type(type = "jsonb")
+    @Column(columnDefinition = "jsonb")
+    private String changed;
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply console/src/main/java/org/georchestra/console/model/AdminLogEntry.java]: > json type = fast insert, no index (slow querying in the end)

Be carrefull with `jsonb`, Postgres seems to be not able to calculate the correct statistique plan with this type. Some requests could be slow with `jsonb` and need some actions to be optimize.

With `changes` data, i thing this is not a real issue.

File: console/pom.xml (hunk 1 above)
Comment (RemiDesgrange) [reply console/pom.xml]: ok it's here : https://github.com/vladmihalcea/hibernate-types#setting-the-hibernatetypesprintbannerfalse-configuration-setting

Comment (Gaetanbrl) [reply console/pom.xml]: Nice catch.

**Comment by Gaetanbrl** [comment]
> USER_ATTRIBUTE_CHANGED logged for title, facsimileTelephoneNumber and telephoneNumber (wtf?)

Weird behavior. If i only change the "title", i get a `java.lang.NullPointerException`.
But the data seems great : 

`{"new":"test","field":"title","old":"","type":"Attribute was changed for a user"}`

This exception seems to arrived with  `logDao.save(log)` : 

https://github.com/jdev-org/georchestra/blob/281dcfa982d69326b15a03c79cc92d1ff21cc956/console/src/main/java/org/georchestra/console/ws/utils/LogUtils.java#L61

**Comment by Gaetanbrl** [comment]
> set -> USER_ATTRIBUTE_CHANGED logged for title, facsimileTelephoneNumber and telephoneNumber (wtf?)

Fix apply & tested. Work better. I commit this fix after other fix.

**Comment by Gaetanbrl** [comment]
> logged, details show Replace the value by . -> should show amount of comms in dept ? 

@landryb  Previous choices was to display length (according to a limit) or list of cities . Indeed, logs details message is wrong.

File: console/src/main/java/org/georchestra/console/model/AdminLogEntry.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (RemiDesgrange) [reply console/src/main/java/org/georchestra/console/model/AdminLogEntry.java]: In french but still : https://blog.anayrat.info/2017/11/26/postgresql-jsonb-et-statistiques/

**Comment by pierrejego** [comment]
> Changes adopted according to reviews :
> 
> **Front**
> 
> * [x]  Home, user, logs templates to a "logger" component (no dupplication, move logic from template to component)
> * [x]  Most uniform code for chosen list insertion (logs tab)
> * [ ]  ~Use another approach UI to display logs details~
> 
> **Back**
> 
> * [x]  Change private nature if necessary
> * [x]  Check dead code
> * [x]  Move logs creation after update or modification
> * [x]  Save log changes information as jsonb instead of @lob
> * [x]  Move create logs into controller instead of dao
> * [x]  Move some methods from controller to logUtils class
> * [ ]  ~Use ogc-server-statistics"-like approach~
> * [x ]  Avoid series of "if" statement

**Comment by pierrejego** [comment]
@Gaetanbrl  add lastest fix and test again :)

**Comment by Gaetanbrl** [comment]
- [ ] create new pending org -> PENDING_ORG_CREATED logged
- [ ] create new pending user -> PENDING_USER_CREATED logged
- [ ] modify org address -> ORG_ATTRIBUTE_CHANGED(address) logged
- [ ] remove pending org -> PENDING_ORG_REFUSED logged
- [ ] remove pending user -> PENDING_USER_REFUSED logged
- [ ] validate pending org -> PENDING_ORG_ACCEPTED logged
- [ ] validate pending user -> PENDING_USER_ACCEPTED logged
- [ ] user modifiers his profile via console/account/userdetails, setting telephone number which wasnt set -> USER_ATTRIBUTE_CHANGED logged for title, facsimileTelephoneNumber and telephoneNumber (wtf?). ==> **FIXED & TESTED**
- [ ] admin modifies user description -> USER_ATTRIBUTE_CHANGED(description) logged
- [ ] admin adds user to org -> USER_ATTRIBUTE_CHANGED (org) logged
- [ ] admin removes user from org -> USER_ATTRIBUTE_CHANGED (org) logged
- [ ] admin adds user to role -> CUSTOM_ROLE_ADDED (role) logged 
- [ ] admin removes user from role DRIVE -> CUSTOM_ROLE_REMOVED (role) logged
- [ ] admin creates role TEST -> ROLE_CREATED logged
- [ ] admin deletes role TEST -> ROLE_DELETED logged
- [ ] admin directly creates org -> ORG_CREATED logged
- [ ] admin deletes org -> ORG_DELETED logged
- [ ] admin directly creates user -> USER_CREATED logged
- [ ]  admin directly deletes user -> USER_DELETED logged
- [ ] user modifiers his password via /console/account/changePassword -> USER_PASSWORD_CHANGED logged
- [ ] admin modifies org area, changing from 1 to complete dept -> ORG_ATTRIBUTE_CHANGED (cities) logged
- [ ] test if details don't display `Replace the value by .` -> `803 élément(s) ajouté(s) à l'emprise de compétence` logged now after fix
- [ ]   admin modifies org area, changing from complete dept to 1 -> ORG_ATTRIBUTE_CHANGED (cities) logged
- [ ] Test if details don't display `Replace the value by .` -> `803 élément(s) supprimé(s) de l'emprise de compétence` logged

**Comment by Gaetanbrl** [comment]
> @Gaetanbrl add lastest fix and test again :)

Tested. Commits will comming soon...

**Comment by landryb** [comment]
with latest commits:

* [x] admin adds 2 comms to org area -> details show `2 élément(s) ajouté(s) à l'emprise de compétence :
[03033, 03230] `
* [x] admin removes 2 comms from org area -> details show `2 élément(s) supprimé(s) de l'emprise de compétence :
[03033, 03230] `
* [x] new validate user logs in, modifies his own phone number -> `USER_ATTRIBUTE_CHANGED (telephoneNumber)`  logged, details show ``` Modification de l'attribut telephoneNumber.

Ajout d'une valeur 0123456789.```

**Comment by pierrejego** [comment]
@fvanderbiest ready for review

**Comment by fvanderbiest** [comment]
Yup, @cmangeat for the server side and @tonio for the client side will have a final look. 
Thanks again for this huge contribution.

File: console/src/main/webapp/manager/app/components/logger/logger.tpl.html [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            <!--refused or deleted-->
+            <span ng-if="logger.getType(log) === ''">{{::log.target}}</span>
+            </td>
+            <!--Type-->
+            <td ng-if="!log.changed"> {{::log.type}}
+            <td>
+            <a ng-if="log.changed && log.changed.field != log.target" title="" ng-click="logger.openLog(log)" href="javascript:void(0);">
+                {{::log.type + (log.changed.field ? ' (' + log.changed.field + ')' : '')}}
+            </a>
+            <span ng-if="log.changed && log.changed.field === log.target">
```
Comment (tonio) [inline console/src/main/webapp/manager/app/components/logger/logger.tpl.html]: In which case `changed.field === target`? I must be missing something here (-:

File: console/src/main/webapp/manager/app/components/logs/logs.tpl.html [hunk 4]
```diff
@@ -1,52 +1,3 @@
...
-        <td><a ng-link="user({id: log.admin, tab: 'infos'})">{{::log.admin}}</a>
-        <td><a ng-link="user({id: log.target, tab: 'infos'})">{{::log.target}}</a>
-        <td>{{::log.type}}
-      </tr>
-    </tbody>
-  </table>
-
-  <dir-pagination-controls></dir-pagination-controls>
-
+  <logger filter="true"></logger>
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply console/src/main/webapp/manager/app/components/logs/logs.tpl.html]: This is to respect the logic of stats component. We have Analytics an stats inside. I think it's a good flexible approach to be able to change logs tab page without impact logger component.

File: console/src/main/webapp/manager/app/components/logger/logger.tpl.html (hunk 3 above)
Comment (Gaetanbrl) [reply console/src/main/webapp/manager/app/components/logger/logger.tpl.html]: Indeed... don't remember this specific case. I thing it was usefull before lasts code modifications... (=> deadcode now ?).

File: console/src/main/webapp/manager/app/components/logger/logger.tpl.html [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            <span ng-if="logger.log.changed.new">: 
+            <br> {{::logger.log.changed.new}} <br>
+            </span>
+            <span ng-if="logger.log.changed.removed"> 
+            <br>{{::logger.log.changed.removed}}
+            </span>
+            <span ng-if="logger.log.changed.removed" translate>logs.removed </span>
+            <span ng-if="logger.log.changed.old">:
+            </br>{{::logger.log.changed.old}}
+            </span>
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply console/src/main/webapp/manager/app/components/logger/logger.tpl.html]: You mean directly create  and return html from JS script ?

**Comment by tonio** [comment]
Data looks incoherent when adding role (custom vs system):
The added role appears in the `new` prop for the first & `old` for the latter.

Screenshot from 2020-01-17 13-31-53

**Comment by Gaetanbrl** [comment]
Nice catch. Fix is ready. I will check your PR first @tonio

File: console/src/main/webapp/manager/app/components/logger/logger.es6 [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+      typeQuery = 'UserLogs'
+    }
+
+    this.logs = this.$injector.get(typeQuery).query(params, () => {
+      // transform each logs changed value into json to find info during html construction
+      this.logs.map(l => {
+        l.changed = l.changed && l.changed.length ? JSON.parse(l.changed) : l.changed
+      })
+      const extract = (key) => [...new Set(this.logs.map(l => l[key]))]
+      this.senders = extract('admin')
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply console/src/main/webapp/manager/app/components/logger/logger.es6]: Wich line(s) are concerned ?

File: console/src/main/webapp/manager/app/components/logs/logs.tpl.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (tonio) [reply console/src/main/webapp/manager/app/components/logs/logs.tpl.html]: good point.

File: console/src/main/webapp/manager/app/components/logger/logger.tpl.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (tonio) [reply console/src/main/webapp/manager/app/components/logger/logger.tpl.html]: It's mainly a string w/o html. I did it in https://github.com/jdev-org/georchestra/pull/4

**Comment by Gaetanbrl** [comment]
@landryb what do you think about jdev-org/georchestra#4 ?

File: console/src/main/webapp/manager/app/components/user/user.es6 [hunk 7]
```diff
@@ -31,16 +32,22 @@ class UserController {
...
         user.validOrg = true
       }
+      // get orgs infos
+      $injector.get('Orgs').query(orgs => {
+        this.orgs = orgs.filter(o => !o.pending)
+        this.orgsId = {}
+        this.orgs.map(org => {
+          this.orgsId[org.id] = org.name
+        })
+      })
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply console/src/main/webapp/manager/app/components/user/user.es6]: Roll back to original version realized.

File: console/src/main/webapp/manager/app/components/user/user.es6 [hunk 8]
```diff
@@ -21,6 +21,7 @@ class UserController {
 
     this.tabs = ['infos', 'roles', 'analytics', 'messages', 'logs', 'manage']
     this.tab = $routeParams.tab
+    this.uid = $routeParams.id
 
     this.user = User.get({ id: $routeParams.id }, (user) => {
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply console/src/main/webapp/manager/app/components/user/user.es6]: Done.

File: console/src/main/webapp/manager/app/components/logger/logger.tpl.html (hunk 3 above)
Comment (Gaetanbrl) [reply console/src/main/webapp/manager/app/components/logger/logger.tpl.html]: Removed this useless case.

**Comment by landryb** [comment]
> @landryb what do you think about [jdev-org#4](https://github.com/jdev-org/georchestra/pull/4) ?

i like the UI approch, fine with me

**Comment by Gaetanbrl** [comment]
@landryb @fvanderbiest here full list of icon to represent admin log type :

image

**We could get more information on action hover :** 

image

image

image

image

**Comment by Gaetanbrl** [comment]
need more tests (specific with logo action).

**Comment by Gaetanbrl** [comment]
As I said - Need more test => To finish correctly :)

**Comment by Gaetanbrl** [comment]
> missing comma

Wich line please ?

**Comment by Gaetanbrl** [comment]
> actually after looking at the context, i think it should be '_added to user_' and '_removed from user_' since the username is put next to it, and '_custom role FOO added to user bar_'/'_system role FOO removed from user bar_' reads better than '_custom role FOO added for this user bar_'/'_system role FOO removed for this user bar_'
> 
> engrish, do you spik it ? ;)

Not cool ! :) I will check traductions ;)

**Comment by landryb** [comment]
In some cases, hovering over an action label for the first time showed `true` for the tooltip, then if a second later i hover again on the same label it shows the correct label in the tooltip. caching issue ? lookup issue ?

**Comment by Gaetanbrl** [comment]
> In some cases, hovering over an action label for the first time showed true for the tooltip, then if a second later i hover again on the same label it shows the correct label in the tooltip. caching issue ? lookup issue ?

Already catched in the UI 1.0.

If you over on html parent, you get "true", else you get info.

**Comment by Gaetanbrl** [comment]
Unless you have no info, you get "true" in some case... not good!

**Comment by Gaetanbrl** [comment]
I hope you have a spanish speaker ;)

**Comment by landryb** [comment]
Fwiw, i think i've tested lots of cases, i'm struggling to see the console in the fr locale (cf #2890) but for me the PR is +1 at this point.

**Comment by fvanderbiest** [comment]
> I hope you have a spanish speaker ;)

We do :-) ping @severo 

One last thing Gaetan, can you import all translations into the spanish and other translation files (even if the translations are not done - ie in english) : this will allow translators to do their job. Thanks.
I usually use `meld` to this.

**Comment by Gaetanbrl** [comment]
> can you import all translations into the spanish and other translation files 

Yes.

> I usually use meld to do this.

New for me, but seems great !

**Comment by Gaetanbrl** [comment]
@fvanderbiest it's good for you (last commits) ?

**Comment by fvanderbiest** [comment]
This looks good. Thanks.
If you're done with it, I'll add translations as far as I can go and then let's merge \o/

**Comment by Gaetanbrl** [comment]
> and then let's merge

😵 😵 \o/

**Pull request merged**